enhancements/olm-deletes-operands.md

```diff
@@ -0,0 +1,209 @@
+
+## Proposal
+
+### User Stories
+
+#### Story 1
+
+As an operator user I want my operator uninstall to also clean up operator-created resources so that I don't have to manually clean up any leftover resources.
+
+### Implementation Details/Notes/Constraints
+
+#### Where to configure cleanup: CSV and/or Subscription
+
+The API to configure operand cleanup could be specified on either the Subscription or ClusterServiceVersion (CSV).
+
+If the cleanup API is only present on the Subscription then operators installed without Subscriptions (by manually creating the CSV and fulfilling requirements) can not opt-in to operand cleanup.
+
+The CSV is the more appropriate point to configure the cleanup option for an operator as the CSV represents the actual installation of the operator, whereas the Subscription only represents the intent to subscribe to updates from the catalog. Moreover, since deleting the CSV causes the operator to be removed, deleting the CSV should be the trigger to initiate operand cleanup. See the [section on CSV finalizers][finalizers-on-csvs] for more details.
+
+#### Deleting CRs
+
+The operand cleanup process, simply put, is the removal of all CRs being reconciled by an operator so it can run finalizers for each CR and clean up any associated operator-created resources.
+
+From OLM's perspective these would be all CRs in the `OperatorGroup`'s [target namespaces][target-namespaces] that are of the same type as the [owned APIservices][owned-apiservices] listed in the CSV. This represents the set of all CRs that the operator reconciles and is responsible for cleaning up after.
+
+OLM should not delete CRs of the type [required APIService][required-apiservices] as those are likely created by the operator, and whose cleanup should be handled by the operator's finalizer.
+
+#### Finalizers on CSVs
+
+OLM can initiate the process of operand clean up when the operator is uninstalled, which is when the CSV is deleted. However, deleting a CSV causes the operator deployment (owned by the CSV) to be garbage collected.
+
+So in order to keep the operator around to process CR deletion events, OLM can add a finalizer to the CSV to prevent the CSV and the operator from being removed. Once OLM has completed operand cleanup and can verify the removal of all CRs, the CSV finalizer can be cleared so that the operator can be safely removed.
+
+```YAML
+apiVersion: operators.coreos.com/v1alpha1
+kind: ClusterServiceVersion
+...
+metadata:
+  ...
+  finalizers:
+  - olm/operand-cleanup
+```
+
+In the event of operand cleanup being stalled due to some operator error, OLM will not clear the CSV finalizer so that the operator stays alive indefinitely. This way the operator can be observed to figure out the required action to unblock cleanup.
+
+#### Operand cleanup API
+
+The CSV spec can have an optional field `spec.cleanup` which lets a user configure the cleanup behaviour before the CSV is deleted.
```

Comment:
I think we also want to bake in the ability to set a deadline on the cleanup, after which the operator's deployment is removed regardless of its cleanup's outcome (or at least show a "TTL exceeded" status condition for users to know something's weird).

Reply:
So my understanding from our previous threads on this is that we decided to not add timeouts (baked in or user-specified) for now. Since that turns the operand cleanup into an imperative action. Plus removing the operator is not really going to help in debugging/resolving the cleanup process if the operator is removed.

See: https://github.com/operator-framework/enhancements/pull/46#discussion_r511038954 and https://github.com/operator-framework/enhancements/pull/46#discussion_r483579433

As for signaling a TTL exceeded on the status, I'm not sure what the default timeout would be that is useful for most applications. There were suggestions ranging from 30s to 15mins.

Also the operator users can still infer if a cleanup is stalled is by looking at the list of pending CRs. But if you think there's value in having a TTL exceeded type condition I can add it to the status if we can define a "reasonable" timeout.